### PR TITLE
[BugFix] Fix the bug of mixed use of current user and qualifiedUser

### LINKED ISSUE #32896: mixed use of current user and qualifiedUser
Because of the mixed use of current user and qualifiedUser, qualifiedUser was used when establishing the connection, and qualifiedUser was modified after executing execute as statement, resulting in the inability to log out the connection record of the corresponding user, resulting in the inability to control max connections.

### PR DESCRIPTION
**Pull request by HangyuanLiu**
Fixes  #32896

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 19 / 61 (31.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/proc/AuthProcDir.java|0|1|00.00%|[76]|
|:large_blue_circle:|com/starrocks/authentication/LDAPAuthProviderForNative.java|0|1|00.00%|[64]|
|:large_blue_circle:|com/starrocks/authentication/KerberosAuthenticationProvider.java|0|1|00.00%|[73]|
|:large_blue_circle:|com/starrocks/http/action/WebBaseAction.java|0|1|00.00%|[236]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|0|1|00.00%|[39]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DropStmtAnalyzer.java|0|4|00.00%|[128, 129, 133, 134]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|1|00.00%|[3655]|
|:large_blue_circle:|com/starrocks/authentication/PlainPasswordAuthenticationProvider.java|0|1|00.00%|[110]|
|:large_blue_circle:|com/starrocks/mysql/privilege/UserPropertyMgr.java|0|10|00.00%|[79, 84, 88, 93, 102, 103, 149, 152, 157, 160]|
|:large_blue_circle:|com/starrocks/privilege/ranger/starrocks/RangerStarRocksAccessController.java|0|1|00.00%|[53]|
|:large_blue_circle:|com/starrocks/privilege/ranger/RangerStarRocksAccessRequest.java|0|2|00.00%|[38, 45]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|0|2|00.00%|[1811, 1812]|
|:large_blue_circle:|com/starrocks/mysql/privilege/PrivTable.java|1|3|33.33%|[231, 314]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|5|15|33.33%|[210, 238, 276, 289, 302, 315, 729, 763, 776, 1389]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|2|50.00%|[185]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationMgr.java|7|10|70.00%|[305, 527, 531]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AuthorizerStmtVisitor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/statistic/StatisticUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/UserIdentity.java|1|1|100.00%|[]|

**Comment by nshangyiming** [comment]
can you give a brief intro to the bug you want to fix and how you fix it?

File: fe/fe-core/src/main/java/com/starrocks/qe/ConnectContext.java [hunk 1]
```diff
@@ -133,14 +133,18 @@ public class ConnectContext {
...
+    // `qualifiedUser` is the user used when the user establishes connection and authentication.
+    // It is the real user used for this connection.
+    // Different from the `currentUserIdentity` authentication user of execute as,
+    // `qualifiedUser` should not be changed during the entire session.
     protected String qualifiedUser;
-    // username@host combination for the StarRocks account
-    // that the server used to authenticate the current client.
-    // In other word, currentUserIdentity is the entry that matched in StarRocks auth table.
-    // This account determines user's access privileges.
+    // `currentUserIdentity` is the user used for authentication. Under normal circumstances,
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/ConnectContext.java:141]:
```suggestion
    // `currentUserIdentity` is the user used for authorization. Under normal circumstances,
```

File: fe/fe-core/src/main/java/com/starrocks/qe/ConnectContext.java [hunk 2]
```diff
@@ -133,14 +133,18 @@ public class ConnectContext {
...
+    // `currentUserIdentity` and `qualifiedUser` are the same user,
+    // but currentUserIdentity may be modified by execute as statement.
     protected UserIdentity currentUserIdentity;
+    // currentRoleIds is the role that has taken effect in the current session.
+    // Note that this set is not all roles belonging to the current user.
+    // For specific logic, please refer to setCurrentRoleIds.
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/ConnectContext.java:147]: how about adding: after a execute as b, the current role ids change to b's

**Review by nshangyiming — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 19 / 61 (31.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/proc/AuthProcDir.java|0|1|00.00%|[76]|
|:large_blue_circle:|com/starrocks/authentication/LDAPAuthProviderForNative.java|0|1|00.00%|[64]|
|:large_blue_circle:|com/starrocks/authentication/KerberosAuthenticationProvider.java|0|1|00.00%|[73]|
|:large_blue_circle:|com/starrocks/http/action/WebBaseAction.java|0|1|00.00%|[236]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|0|1|00.00%|[39]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DropStmtAnalyzer.java|0|4|00.00%|[128, 129, 133, 134]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|1|00.00%|[3655]|
|:large_blue_circle:|com/starrocks/authentication/PlainPasswordAuthenticationProvider.java|0|1|00.00%|[110]|
|:large_blue_circle:|com/starrocks/mysql/privilege/UserPropertyMgr.java|0|10|00.00%|[79, 84, 88, 93, 102, 103, 149, 152, 157, 160]|
|:large_blue_circle:|com/starrocks/privilege/ranger/starrocks/RangerStarRocksAccessController.java|0|1|00.00%|[53]|
|:large_blue_circle:|com/starrocks/privilege/ranger/RangerStarRocksAccessRequest.java|0|2|00.00%|[38, 45]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|0|2|00.00%|[1811, 1812]|
|:large_blue_circle:|com/starrocks/mysql/privilege/PrivTable.java|1|3|33.33%|[231, 314]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|5|15|33.33%|[210, 238, 276, 289, 302, 315, 729, 763, 776, 1389]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|2|50.00%|[185]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationMgr.java|7|10|70.00%|[305, 527, 531]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AuthorizerStmtVisitor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/statistic/StatisticUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/UserIdentity.java|1|1|100.00%|[]|

**Review by Astralidea — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**